metric_learn/mlkr.py

```diff
@@ -79,34 +87,68 @@ def fit(self, X, y):
       """
       X, y, A = self._process_inputs(X, y)
 
+      # Measure the total training time
+      t_train = time.time()
+
       # note: this line takes (n*n*d) memory!
       # for larger datasets, we'll need to compute dX as we go
       dX = (X[None] - X[:, None]).reshape((-1, X.shape[1]))
 
-      res = minimize(_loss, A.ravel(), (X, y, dX), method='CG', jac=True,
+      self.n_iter_ = 0
+      res = minimize(self._loss, A.ravel(), (X, y, dX), method='CG', jac=True,
                      tol=self.alpha,
                      options=dict(maxiter=self.max_iter, eps=self.epsilon))
       self.transformer_ = res.x.reshape(A.shape)
-      self.n_iter_ = res.nit
+
+      if self.verbose:
+          cls_name = self.__class__.__name__
+          # Warn the user if the algorithm did not converge
+          if not res.success:
+              warnings.warn('[{}] MLKR did not converge: {}'
+                            .format(cls_name, res.message), ConvergenceWarning)
+          print('[{}] Training took {:8.2f}s.'.format(cls_name, t_train))
+
       return self
 
   def transformer(self):
       return self.transformer_
 
-
-def _loss(flatA, X, y, dX):
-  A = flatA.reshape((-1, X.shape[1]))
-  dist = pdist(X, metric='mahalanobis', VI=A.T.dot(A))
-  K = squareform(np.exp(-dist**2))
-  denom = np.maximum(K.sum(axis=0), EPS)
-  yhat = K.dot(y) / denom
-  ydiff = yhat - y
-  cost = (ydiff**2).sum()
-
-  # also compute the gradient
-  np.fill_diagonal(K, 1)
-  W = 2 * K * (np.outer(ydiff, ydiff) / denom)
-  # note: this is the part that the matlab impl drops to C for
-  M = (dX.T * W.ravel()).dot(dX)
-  grad = 2 * A.dot(M)
-  return cost, grad.ravel()
+  def _loss(self, flatA, X, y, dX):
+
+    if self.n_iter_ == 0:
+      if self.verbose:
+        header_fields = ['Iteration', 'Objective Value', 'Time(s)']
+        header_fmt = '{:>10} {:>20} {:>10}'
+        header = header_fmt.format(*header_fields)
+        cls_name = self.__class__.__name__
+        print('[{}]'.format(cls_name))
+        print('[{}] {}\n[{}] {}'.format(cls_name, header, cls_name,
+                                        '-' * len(header)))
```

Comment:
Named format slots might make this easier to read. That is, like the "by name" examples here: https://docs.python.org/2/library/string.html#format-examples

Reply:
Agreed, done (except for the header field because there I thought keyword arguments would be more redundant than really helping reading , but tell me if you don't think so)